app/src/test/java/supermarket/ProductTest.kt

```diff
@@ -1,84 +1,303 @@
 package supermarket
 
-import org.hamcrest.CoreMatchers.`is` as Is
+import org.hamcrest.CoreMatchers.containsString
 import org.hamcrest.MatcherAssert.assertThat
-import org.hamcrest.Matchers.closeTo
+import org.junit.After
+import org.junit.Assert.fail
 import org.junit.Before
 import org.junit.Test
 import supermarket.TestUtils.apples
+import supermarket.TestUtils.flour
+import supermarket.TestUtils.oranges
 import supermarket.TestUtils.toothbrush
 import supermarket.TestUtils.toothpaste
 import supermarket.model.BundleOffer
 import supermarket.model.PercentageOffer
+import supermarket.model.QuantityForAmountOffer
+import supermarket.model.Receipt
 import supermarket.model.ShoppingCart
 import supermarket.model.Teller
 import supermarket.model.ThreeForTwoOffer
+import org.hamcrest.CoreMatchers.`is` as Is
 
 class ProductTest {
     private val catalog = FakeCatalog()
     private val cart = ShoppingCart()
     private val teller = Teller(catalog)
+    private var receipt: Receipt? = null
 
     @Before
     fun setup() {
         TestUtils.setupCatalog(catalog)
     }
 
-    @Test
-    fun `toothbrush discount is recognized`() {
-        cart.addItemQuantity(toothbrush, 1.0)
+    @After
+    fun tearDown() {
+        // No teardown necessary, but print receipt while we're at it.
+        receipt?.let {
+            println(ReceiptPrinter().printReceipt(receipt!!))
+        }
+    }
 
+    @Test
+    fun `Require items in cart to apply discounts`() {
         teller.addSpecialOffer(PercentageOffer(toothbrush, 10.0))
 
-        val receipt = teller.checksOutArticlesFrom(cart)
-        val discounts = receipt.getDiscounts()
-        assertThat(discounts.size, Is(1))
+        receipt = teller.checksOutArticlesFrom(cart)
+        val discounts = receipt!!.getDiscounts()
```

Comment:
I dig where you're going with the printing, but I don't think it's worth the `!!` everywhere.  Maybe just a printReceipt() function in this class which would let individual tests print if they want, which would DRY up the ReceiptPrinter calls at least?

Reply:
I could lateinit instead and just check that it's initialized in the tearDown function?